### ISSUE TITLE
[RFC] Mistral v2.0 Roadmap

### ISSUE THREAD

**Issue by J38** [stanford-crfm/mistral#109]
I had some thoughts of what to add to Mistral based on my experiences of training a domain specific model.

* facilitate using local/custom data sets in `*.jsonl` files

* facilitate data blends with loading script

* add data filtering to data process so worst N% examples can be removed via model X

* add tests for data processing

* tokenization crashes when using deepspeed ... refactor entry point to run data processing and then call launcher with subprocess

* new CLI allowing for just running data processing and specifying launcher

```
mistral process-data --config conf/gpt2-small-demo.yaml
mistral train --config conf/gpt2-small-demo.yaml --launcher deepspeed
```

* logging/reporting/inspection of dataset(s) used for model, stats on dataset, ability to display samples

* update to latest version of dependencies, update docs on how to build the environment

* clean up configs/resources

* do a new pass on documentation

Please let me know what you think, and feel free to add ideas of things to add. Thanks!

**Comment by J38** [comment]
For instance, I found it annoying that I had to run this fake run of my training process to handle the data processing, then launch again with deepspeed. So either I'm doing something wrong and we can run the data processing with deepspeed or we should refactor this to resolve that issue.

**Comment by J38** [comment]
And it seems like it'd nice if you could just specify the launcher you want (e.g. deepspeed or None or torch.distributed.launch) and have it just work instead of having to write out that command.

**Comment by J38** [comment]
Also I found it annoying to not have a progress bar during the first step of data loading when I was dealing with 100GB of data ... not sure if we could try to do something about that .

**Comment by dlwh** [comment]
> logging/reporting/inspection of dataset(s) used for model, stats on dataset, ability to display samples

For that I think we should start with trainer.create_model_card which does a lot of that.

Displaying samples is also something HF's model infrastructure does well and we should piggyback on that.

**Comment by dlwh** [comment]
> I had some thoughts of what to add to Mistral based on my experiences of training a domain specific model. Some of these things have already been implemented.
> 
>     facilitate using local/custom data sets in *.jsonl files (done)

I would prefer if we moved to supporting loading a bundle of HF-datasets native things. It's not too much work to write a dataset script, and it makes creating model cards etc. easier.

> 
>     facilitate data blends with loading script (done)

This is an import thing.

> 
>     add data filtering to data process so worst N% examples according to model X can be removed

💯 

> 
>     add tests for data processing

💯 

> 
>     tokenization crashes when using deepspeed ... refactor entry point to run data processing and then call launcher with subprocess after data has been processed

I hate python. But yes I agree this is a good idea.

> 
>     new CLI allowing for just running data processing and specifying launcher
> 
> mistral process-data --config conf/gpt2-small-demo.yaml
> mistral train --config conf/gpt2-small-demo.yaml --launcher deepspeed # if necessary this would run data processing

This seems better than the weird inside-out deepspeed thing that HF does.

> 
>     logging/reporting/inspection of dataset(s) used for model, stats on dataset, ability to display samples

as i said above, we should start with just supporting model cards and see how we feel.

> 
>     update to latest version of dependencies, update docs on how to build the environment
> 
>     clean up configs/resources
> 
>     do a new pass on documentation
> 
>     update documentation on accessing new Mistral checkpoints on HF Hub
> 
>     update Mistral checkpoints so main branch points to checkpoint-400000
> 
>     fix issue where eval dataset is hard coded to be "openwebtext_ppl"
> 
> Please let me know what you think, and feel free to add ideas of things to add. Thanks!

**Comment by dlwh** [comment]
As I said on Slack, I'm now more bullish on us just supporting loading a checkpoint and fine tuning. It's like 3 lines of code, doesn't change any other code paths (I think??) and shouldn't impose any further maintenance. Famous last words, I know, but it seems

**Comment by dlwh** [comment]
how do y'all feel about kanbans? It's kind of the only project management tool I like.

**Comment by siddk** [comment]
Big fan - we can use the Github Project ones?

**Comment by dlwh** [comment]
already made! https://github.com/orgs/stanford-crfm/projects/1?add_cards_query=is%3Aopen (It's empty, and, for the moment, private to crfm folks, for any watchers)

**Comment by J38** [comment]
Also this is a bit off topic, but back in December there was some work done identifying some issues with DeepSpeed + perfect reproducibility ... basically some of the ways DeepSpeed initializes from checkpoint violate perfect reproducibility ... the way the scheduler and learning rate get set up from a checkpoint are off a bit in the DeepSpeed codebase ... if we still wanted to push on perfect reproducibility then we might want to put something together for them to look at .... by "we" I mean me :(

**Comment by dlwh** [comment]
i think we should revisit OnlineBenchmarkTrainer and see if we can stop using it, in favor of callbacks. It's a bit fragile as is in terms of HF updates.

**Comment by dlwh** [comment]
everything is done here except for checkpoints
